src/main/java/org/elasticsearch/indices/SyncedFlushService.java

```diff
@@ -0,0 +1,755 @@
+        }
+    }
+
+    final IndexShardRoutingTable getActiveShardRoutings(ShardId shardId, ClusterState state) {
+        final IndexRoutingTable indexRoutingTable = state.routingTable().index(shardId.index().name());
+        if (indexRoutingTable == null) {
+            IndexMetaData index = state.getMetaData().index(shardId.index().getName());
+            if (index != null && index.state() == IndexMetaData.State.CLOSE) {
+                throw new IndexClosedException(shardId.index());
+            }
+            throw new IndexMissingException(shardId.index());
+        }
+        final IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(shardId.id());
+        if (shardRoutingTable == null) {
+            throw new IndexShardMissingException(shardId);
+        }
+        return shardRoutingTable;
+    }
+
+    /**
+     * returns the number of inflight operations on primary. -1 upon error.
+     */
+    protected int getInflightOpsCount(final ShardId shardId, ClusterState state, IndexShardRoutingTable shardRoutingTable) {
+        final ShardRouting primaryShard = shardRoutingTable.primaryShard();
+        final DiscoveryNode primaryNode = state.nodes().get(primaryShard.currentNodeId());
+        if (primaryNode == null) {
+            logger.trace("{} failed to resolve node for primary shard {}, skipping sync", shardId, primaryShard);
+            return -1;
+        }
+        final AtomicInteger result = new AtomicInteger(-1);
+        final CountDownLatch latch = new CountDownLatch(1);
+        logger.trace("{} retrieving in flight operation count", shardId);
+        transportService.sendRequest(primaryNode, IN_FLIGHT_OPS_ACTION_NAME, new InFlightOpsRequest(shardId),
+                new BaseTransportResponseHandler<InFlightOpsResponse>() {
+                    @Override
+                    public InFlightOpsResponse newInstance() {
+                        return new InFlightOpsResponse();
+                    }
+
+                    @Override
+                    public void handleResponse(InFlightOpsResponse response) {
+                        result.set(response.opCount());
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        logger.debug("{} unexpected error while retrieving inflight op count", shardId);
+                        result.set(-1);
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public String executor() {
+                        return ThreadPool.Names.SAME;
+                    }
+                });
+        try {
+            if (latch.await(inflightOpsTimeout.millis(), TimeUnit.MILLISECONDS) == false) {
+                logger.debug("{} in flight operation check timed out after [{}]", shardId, syncTimeout);
+            }
+        } catch (InterruptedException e) {
```

Comment:
we have to restore the interrupt status here with `Thread.currentThread.interrupt()`